cmd/src/campaign_progress_printer_test.go

```diff
@@ -0,0 +1,249 @@
+	want := []string{
+		"⠋  Executing... (0/3, 0 errored)  ",
+		"│                                                                               ",
+		"├── github.com/sourcegraph/sourcegraph  echo Hello World > README.md          0s",
+		"├── github.com/sourcegraph/src-cli      Downloading archive                   0s",
+		"└── github.com/sourcegraph/automati...  echo Hello World > README.md          0s",
+	}
+	if !cmp.Equal(want, have) {
+		t.Fatalf("wrong output:\n%s", cmp.Diff(want, have))
+	}
+
+	// Now mark the last task as completed
+	statuses[len(statuses)-1] = &campaigns.TaskStatus{
+		RepoName:           "github.com/sourcegraph/automation-testing",
+		StartedAt:          time.Now().Add(time.Duration(-5) * time.Second),
+		FinishedAt:         time.Now().Add(time.Duration(5) * time.Second),
+		CurrentlyExecuting: "",
+		Err:                nil,
+		ChangesetSpec: &campaigns.ChangesetSpec{
+			BaseRepository: "graphql-id",
+			CreatedChangeset: &campaigns.CreatedChangeset{
+				BaseRef:        "refs/heads/main",
+				BaseRev:        "d34db33f",
+				HeadRepository: "graphql-id",
+				HeadRef:        "refs/heads/my-campaign",
+				Title:          "This is my campaign",
+				Body:           "This is my campaign",
+				Commits: []campaigns.GitCommitDescription{
+					{
+						Message: "This is my campaign",
+						Diff:    progressPrinterDiff,
+					},
+				},
+				Published: false,
+			},
+		},
+	}
+
+	printer.PrintStatuses(statuses)
+	have = buf.Lines()
+	want = []string{
+		"github.com/sourcegraph/automation-testing",
+		"\tREADME.md   | 3 +++",
+		"\ta/b/c/c.txt | 1 -",
+		"\tx/x.txt     | 2 +-",
+		"  3 files changed, 4 insertions, 2 deletions",
+		"  Execution took 10s",
+		"",
+		"⠋  Executing... (1/3, 0 errored)  ███████████████▍",
+		"│                                                                                 0s", // Not sure why this is here, bug?
+		"├── github.com/sourcegraph/sourcegraph  echo Hello World > README.md          0s",
+		"├── github.com/sourcegraph/src-cli      Downloading archive                   0s",
+		"└── github.com/sourcegraph/automati...  3 files changed ++++               0s",
+	}
+	if !cmp.Equal(want, have) {
+		t.Fatalf("wrong output:\n%s", cmp.Diff(want, have))
+	}
+
+	// Print again to make sure we get the same result
+	printer.PrintStatuses(statuses)
+	have = buf.Lines()
+	if !cmp.Equal(want, have) {
+		t.Fatalf("wrong output:\n%s", cmp.Diff(want, have))
+	}
+
+}
+
+type ttyBuf struct {
+	lines [][]byte
+
+	line   int
+	column int
+}
+
+func (t *ttyBuf) Write(b []byte) (int, error) {
+	var cur int
+
+	for cur < len(b) {
+		switch b[cur] {
+		case '\n':
+			t.line++
+			t.column = 0
+		case '\x1b':
+			// First of all: forgive me.
+			//
+			// Now. Looks like we ran into a VT100 escape code.
+			// They follow this structure:
+			//
+			//      \x1b [ <digit> <command>
+			//
+			// So we jump over the \x1b[ and try to parse the digit.
+
+			cur = cur + 2 // cur == '\x1b', cur + 1 == '['
```

Comment:
Do you want to verify that `cur+1 == '['`?

Reply:
Done!